src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/BindSymbolEvaluator.cs

```diff
@@ -86,16 +90,22 @@ public async Task EvaluateBindedSymbolsAsync(IEnumerable<BindSymbol> symbols, IV
             var successfulTasks = tasksToRun.Where(t => t.Task.IsCompleted && t.Task.Result != null);
             foreach (var task in successfulTasks)
             {
-                object? value = ParameterConverter.InferTypeAndConvertLiteral(task.Task.Result!);
+                string obtainedValue = task.Task.Result!;
+                BindSymbol currentSymbol = task.Symbol;
 
-                if (value != null)
+                bool result = ParameterConverter.TryConvertLiteralToDatatype(obtainedValue, currentSymbol.DataType, out object? value);
+                if (result && value != null)
+                {
+                    variableCollection[currentSymbol.Name] = value;
+                    _logger.LogDebug("Variable '{0}' was set to '{1}'.", currentSymbol.Name, value);
+                }
+                if (!result && !string.IsNullOrWhiteSpace(currentSymbol.DataType))
                 {
-                    variableCollection[task.Symbol.Name] = value;
-                    _logger.LogDebug("Variable '{0}' was set to '{1}'.", task.Symbol.Name, value);
+                    _logger.LogWarning(LocalizableStrings.BindSymbolEvaluator_Warning_ConversionFailure, currentSymbol.Name, obtainedValue, currentSymbol.DataType);
                 }
                 else
                 {
-                    _logger.LogDebug("Variable '{0}' was not set due to its value is null.", task.Symbol.Name);
+                    _logger.LogDebug("Variable '{0}' was not set due to its value is null.", currentSymbol.Name);
```

Comment:
Should this be something like "Variable '{0}' was not set as its value '{1}' couldn't be converted to requested type '{2}'."?